cmd/zdb/zdb.c

```diff
@@ -5776,41 +5816,117 @@ zdb_count_block(zdb_cb_t *zcb, zilog_t *zilog, const blkptr_t *bp,
 			zbre = umem_zalloc(sizeof (zdb_brt_entry_t),
 			    UMEM_NOFAIL);
 			zbre->zbre_dva = bp->blk_dva[0];
-			zbre->zbre_refcount = crefcnt;
+			zbre->zbre_refcount = refcnt;
 			avl_insert(&zcb->zcb_brt, zbre, where);
 		}
 	}
 
-	if (dump_opt['L'])
-		return;
-
 	if (BP_GET_DEDUP(bp)) {
-		ddt_t *ddt;
-		ddt_entry_t *dde;
+		/*
+		 * Dedup'd blocks are special. We need to count them, so we can
+		 * later uncount them when reporting leaked space, and we must
+		 * only claim them once.
+		 *
+		 * We use the existing dedup system to track what we've seen.
+		 * The first time we see a block, we do a ddt_lookup() to see
+		 * if it exists in the DDT. This function also properly
+		 * resolves any unflushed logs. If we're doing leak tracking,
+		 * we claim the block at this time.
+		 *
+		 * Each time we see a block, we reduce the refcount in the
+		 * entry by one, and add to the size and count of dedup'd
+		 * blocks to report at the end.
+		 */
+
+		ddt_t *ddt = ddt_select(zcb->zcb_spa, bp);
 
-		ddt = ddt_select(zcb->zcb_spa, bp);
 		ddt_enter(ddt);
-		dde = ddt_lookup(ddt, bp, B_FALSE);
+		spa_config_enter(spa, SCL_CONFIG, FTAG, RW_READER);
 
+		/*
+		 * First, we look for the block in the live tree. If it's not
+		 * there, then this is the first time we've seen this block and
+		 * we need to make a note to do some extra work.
+		 */
+		boolean_t first = B_FALSE;
+
+		ddt_key_t search;
+		ddt_key_fill(&search, bp);
+		ddt_entry_t *dde = avl_find(&ddt->ddt_tree, &search, NULL);
+
+		ddt_phys_variant_t v = DDT_PHYS_NONE;
 		if (dde == NULL) {
-			refcnt = 0;
+			/* First time we've seen this block, load it's entry. */
+			dde = ddt_lookup(ddt, bp);
+			if (dde == NULL) {
+				/*
+				 * If ddt_lookup() returns NULL, then this
+				 * block previously had an entry, then was
+				 * pruned, and then a new one has been created
+				 * for a later generation of the block.
+				 *
+				 * In this case, it's already been cleaned up,
+				 * so do nothing.
+				 */
+			}
+			else {
+				/* Got a real entry for it, note this fact. */
+				v = ddt_phys_select(ddt, dde, bp);
+				first = B_TRUE;
+			}
 		} else {
-			ddt_phys_variant_t v = ddt_phys_select(ddt, dde, bp);
-			refcnt = ddt_phys_decref(dde->dde_phys, v);
-			if (ddt_phys_total_refcnt(ddt, dde->dde_phys) == 0)
-				ddt_remove(ddt, dde);
+			/*
+			 * We have seen this block before. Like above, get phys
+			 * variant and check birth time, skip if its for a
+			 * newer version of this block.
+			 */
+			v = ddt_phys_select(ddt, dde, bp);
+		}
+
+		if (dde != NULL && v < DDT_PHYS_NONE) {
+			/* We now have a matching entry */
+
+			/* Consume a reference for this block. */
+			uint64_t refcnt = ddt_phys_refcnt(dde->dde_phys, v);
+			if (refcnt > 0)
+				ddt_phys_decref(dde->dde_phys, v);
+			else {
+				/*
+				 * XXX what do we do here? we saw more valid
+				 *     blocks we have refcounts for. we should
+				 *     probably report _something_
+				 *       -- robn, 2024-06-24
+				*/
```

Comment:
Error needed here?

Reply:
Rob will post an update to this code in the new few days